{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "8cf7ede8",
   "metadata": {},
   "source": [
    "# BIOL 33000: Genomics, Bioinformatics, and Computational Biology.\n",
    "### Computer Lab 01: 20240828"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4ecf77c3",
   "metadata": {},
   "source": [
    "## Introduction and overview"
   ]
  },
  {
   "cell_type": "raw",
   "id": "c23d8e45",
   "metadata": {},
   "source": [
    "Welcome to Genomics Lab!\n",
    "\n",
    "Modern biology generates tons of data, from sequencing projects, to analyses of gene expression, to high-throughput cell-based drug assays, to signatures of different microbiota in our gut or in the environment. Computational skills and data science knowledge are in high demand in many industries, and are becoming nearly mandatory for careers in biology.\n",
    "\n",
    "Coding takes practice, and can be challenging. As upper-level biology & biochemistry students, you have pretty much figured out how to do well in science classes. Coding is likely brand new for most of you, and it can be frustrating to feel like a beginner when you have already advanced so far, and worked so hard. Another challenge is that computers are stupid and unforgiving. They do not tolerate mistakes, and cannot tell the difference between matplotlib.pyplot and matplotlib.pylot.\n",
    "\n",
    "Like learning a musical instrument or a new sport, you will need a lot of practice to become proficient. If you love to solve puzzles or riddles or logic games, I think you'll like to code. If you like to crunch numbers and really get into the data, I think you'll like to code. If you like to create effective and compelling visualizations of scientific concepts, I think you'll like to code! If you don't like any of these things, learning to code will still be good for you."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d59fc95e",
   "metadata": {},
   "source": [
    "## Jupyter notebooks"
   ]
  },
  {
   "cell_type": "raw",
   "id": "150a5ad0",
   "metadata": {},
   "source": [
    "First, some nitty gritty. You are looking at a Jupyter notebook here. These notebooks have several advantages:\n",
    "\n",
    "1. They work on all computer platforms, from Macs to Windows to Linux\n",
    "\n",
    "2. They enable good computer programming practices, especially for beginners. They encourage you to do things one step at a time.\n",
    "\n",
    "3. They are flexible - we are going to use them for Python coding, but you can code in many other programming languages, and you can also do neat things like display graphics and other content. \n",
    "\n",
    "For more about what you can do with Jupyter, see:\n",
    "https://www.dataquest.io/blog/jupyter-notebook-tutorial/\n",
    "\n",
    "A typical Jupyter notebook is built from individual blocks, or cells. These cells can contain plain text (like this one), or formatted text (e.g. for section headings like the thing that says \"Jupyter notebooks\" above), or computer code. "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f76c6894",
   "metadata": {},
   "source": [
    "## A little demo about Jupyter notebooks"
   ]
  },
  {
   "cell_type": "raw",
   "id": "b9d68344",
   "metadata": {},
   "source": [
    "For lots of help, see:\n",
    "https://jupyter-notebook.readthedocs.io/en/stable/examples/Notebook/What%20is%20the%20Jupyter%20Notebook.html\n",
    "\n",
    "If you see a lime-green outline of a cell, you are in 'edit mode', and you can edit the contents of the cell.\n",
    "\n",
    "To change mode from 'edit mode' to 'command mode' (blue outline), press escape. In command mode, you can:\n",
    "    Type 'b' to add a cell below\n",
    "    Type 'a' to add a cell above\n",
    "    Type dd to delete a cell"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f66569ac",
   "metadata": {},
   "source": [
    "## The hardest parts about getting started"
   ]
  },
  {
   "cell_type": "raw",
   "id": "2834e930",
   "metadata": {},
   "source": [
    "There are three hard things about getting started with coding:\n",
    "\n",
    "1. You need to have a super high tolerance for failure and frustration, and you need to be persistent in the face of failure. Failure is part of the process. I would guess that failure is >99% of coding. This is a shift in mindset - you have gone through almost two decades of school pursuing the 'right' answers. In coding, you will be 'wrong' again and again. And again. The key is to just keep going until you finally get it right. \n",
    "\n",
    "2. You need to figure out how to break big things down into tiny little steps, and do the steps one at a time. \n",
    "\n",
    "3. You need to figure out how to navigate your computer in a way that you are not used to ... basically, you need to be sure that everything you need is within a single folder on your computer, and you need to know many different ways to find that folder. Sounds easy, but in my experience, this is the biggest hurdle for beginners."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3225af56",
   "metadata": {},
   "source": [
    "# Introduction to PsychENCODE"
   ]
  },
  {
   "cell_type": "raw",
   "id": "57c129d2",
   "metadata": {},
   "source": [
    "We will start by downloading and examining some data from the PsychENCODE project.\n",
    "\n",
    "Psych = psychiatric conditions (and general brain function and development).\n",
    "ENCODE = ENCyclopedia Of DNA Elements.\n",
    "\n",
    "Basically, this group of researchers wanted to find out everything they could about how the genome regulates brain development and brain function. They also wanted to see what differences in the genome are associated with conditions where brain function is different, such as in Autism Spectrum Disorder, Schizophrenia, and Alzheimer's disease.\n",
    "\n",
    "To do so, they collected brain samples from control individuals, and from individuals with the conditions above. They then did huge number of tests on these samples, from genome sequencing, to gene expression analysis, to assays of epigenetic marks on chromosomes, and more. This approach is often described as an '-omics' analysis.\n",
    "\n",
    "Today we will download some high-level data from this project, and use Python coding (and associated tools) to visualize the data."
   ]
  },
  {
   "cell_type": "raw",
   "id": "809a9385",
   "metadata": {},
   "source": [
    "### working\n",
    "psychencode data home: http://brainscope.psychencode.org/\n",
    "brainscope\n",
    "key resources\n",
    "sample metadata\n",
    "BE SURE TO DOWNLOAD IN SAME FOLDER AS THIS NOTEBOOK!\n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "223ee7b1",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "import numpy as np"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "9db68cc5",
   "metadata": {},
   "outputs": [],
   "source": [
    "# this is a CODE block!\n",
    "# we can 'comment out' stuff that we don't want to run ... with a '#'\n",
    "\n",
    "# reading text file in to a dataframe\n",
    "# see https://pandas.pydata.org/docs/reference/api/pandas.read_csv.html\n",
    "\n",
    "# we are starting with a file that contains a table\n",
    "\n",
    "df = pd.read_csv('PEC2_sample_metadata.txt')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "33bf48d2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Cohort\\tIndividual_ID\\tBiological_Sex\\tAge_death\\tDisorder\\t1000G_ancestry\\tGenotype_data\\tsnATAC-Seq\\tPMI\\tpH\\tRIN\\tNotes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CMC\\tCMC_MSSM_002\\tmale\\t42\\tcontrol\\tEUR\\tArr...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CMC\\tCMC_MSSM_033\\tfemale\\t81\\tcontrol\\tAFR\\tA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CMC\\tCMC_MSSM_035\\tmale\\t75\\tcontrol\\tEUR\\tArr...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Cohort\\tIndividual_ID\\tBiological_Sex\\tAge_death\\tDisorder\\t1000G_ancestry\\tGenotype_data\\tsnATAC-Seq\\tPMI\\tpH\\tRIN\\tNotes\n",
       "0  CMC\\tCMC_MSSM_002\\tmale\\t42\\tcontrol\\tEUR\\tArr...                                                                        \n",
       "1  CMC\\tCMC_MSSM_033\\tfemale\\t81\\tcontrol\\tAFR\\tA...                                                                        \n",
       "2  CMC\\tCMC_MSSM_035\\tmale\\t75\\tcontrol\\tEUR\\tArr...                                                                        "
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "3915a166",
   "metadata": {},
   "outputs": [],
   "source": [
    "# that didn't work! try again. this is a tab-delimited table\n",
    "df = pd.read_csv('PEC2_sample_metadata.txt',sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "7ab8c042",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Cohort</th>\n",
       "      <th>Individual_ID</th>\n",
       "      <th>Biological_Sex</th>\n",
       "      <th>Age_death</th>\n",
       "      <th>Disorder</th>\n",
       "      <th>1000G_ancestry</th>\n",
       "      <th>Genotype_data</th>\n",
       "      <th>snATAC-Seq</th>\n",
       "      <th>PMI</th>\n",
       "      <th>pH</th>\n",
       "      <th>RIN</th>\n",
       "      <th>Notes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CMC</td>\n",
       "      <td>CMC_MSSM_002</td>\n",
       "      <td>male</td>\n",
       "      <td>42</td>\n",
       "      <td>control</td>\n",
       "      <td>EUR</td>\n",
       "      <td>Array</td>\n",
       "      <td>NaN</td>\n",
       "      <td>22.3</td>\n",
       "      <td>6.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CMC</td>\n",
       "      <td>CMC_MSSM_033</td>\n",
       "      <td>female</td>\n",
       "      <td>81</td>\n",
       "      <td>control</td>\n",
       "      <td>AFR</td>\n",
       "      <td>Array</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4.0</td>\n",
       "      <td>6.3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CMC</td>\n",
       "      <td>CMC_MSSM_035</td>\n",
       "      <td>male</td>\n",
       "      <td>75</td>\n",
       "      <td>control</td>\n",
       "      <td>EUR</td>\n",
       "      <td>Array</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CMC</td>\n",
       "      <td>CMC_MSSM_036</td>\n",
       "      <td>female</td>\n",
       "      <td>82</td>\n",
       "      <td>control</td>\n",
       "      <td>EUR</td>\n",
       "      <td>Array</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3.8</td>\n",
       "      <td>6.4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CMC</td>\n",
       "      <td>CMC_MSSM_048</td>\n",
       "      <td>male</td>\n",
       "      <td>84</td>\n",
       "      <td>Schizophrenia</td>\n",
       "      <td>EUR</td>\n",
       "      <td>Array</td>\n",
       "      <td>NaN</td>\n",
       "      <td>17.7</td>\n",
       "      <td>6.7</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>CMC</td>\n",
       "      <td>CMC_MSSM_049</td>\n",
       "      <td>female</td>\n",
       "      <td>61</td>\n",
       "      <td>control</td>\n",
       "      <td>EUR</td>\n",
       "      <td>Array</td>\n",
       "      <td>NaN</td>\n",
       "      <td>20.3</td>\n",
       "      <td>6.9</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>CMC</td>\n",
       "      <td>CMC_MSSM_056</td>\n",
       "      <td>female</td>\n",
       "      <td>67</td>\n",
       "      <td>control</td>\n",
       "      <td>EUR</td>\n",
       "      <td>Array</td>\n",
       "      <td>NaN</td>\n",
       "      <td>16.5</td>\n",
       "      <td>6.7</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>CMC</td>\n",
       "      <td>CMC_MSSM_058</td>\n",
       "      <td>female</td>\n",
       "      <td>76</td>\n",
       "      <td>control</td>\n",
       "      <td>AMR</td>\n",
       "      <td>Array</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4.3</td>\n",
       "      <td>6.5</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>CMC</td>\n",
       "      <td>CMC_MSSM_060</td>\n",
       "      <td>male</td>\n",
       "      <td>89+</td>\n",
       "      <td>control</td>\n",
       "      <td>AFR</td>\n",
       "      <td>Array</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3.2</td>\n",
       "      <td>6.4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>CMC</td>\n",
       "      <td>CMC_MSSM_067</td>\n",
       "      <td>female</td>\n",
       "      <td>81</td>\n",
       "      <td>control</td>\n",
       "      <td>EUR</td>\n",
       "      <td>Array</td>\n",
       "      <td>NaN</td>\n",
       "      <td>19.4</td>\n",
       "      <td>6.4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Cohort Individual_ID Biological_Sex Age_death       Disorder 1000G_ancestry  \\\n",
       "0    CMC  CMC_MSSM_002           male        42        control            EUR   \n",
       "1    CMC  CMC_MSSM_033         female        81        control            AFR   \n",
       "2    CMC  CMC_MSSM_035           male        75        control            EUR   \n",
       "3    CMC  CMC_MSSM_036         female        82        control            EUR   \n",
       "4    CMC  CMC_MSSM_048           male        84  Schizophrenia            EUR   \n",
       "5    CMC  CMC_MSSM_049         female        61        control            EUR   \n",
       "6    CMC  CMC_MSSM_056         female        67        control            EUR   \n",
       "7    CMC  CMC_MSSM_058         female        76        control            AMR   \n",
       "8    CMC  CMC_MSSM_060           male       89+        control            AFR   \n",
       "9    CMC  CMC_MSSM_067         female        81        control            EUR   \n",
       "\n",
       "  Genotype_data snATAC-Seq   PMI   pH  RIN Notes  \n",
       "0         Array        NaN  22.3  6.0  NaN   NaN  \n",
       "1         Array        NaN   4.0  6.3  NaN   NaN  \n",
       "2         Array        NaN   NaN  2.0  NaN   NaN  \n",
       "3         Array        NaN   3.8  6.4  NaN   NaN  \n",
       "4         Array        NaN  17.7  6.7  NaN   NaN  \n",
       "5         Array        NaN  20.3  6.9  NaN   NaN  \n",
       "6         Array        NaN  16.5  6.7  NaN   NaN  \n",
       "7         Array        NaN   4.3  6.5  NaN   NaN  \n",
       "8         Array        NaN   3.2  6.4  NaN   NaN  \n",
       "9         Array        NaN  19.4  6.4  NaN   NaN  "
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.head(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "54f464bb",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'pandas.core.frame.DataFrame'>\n",
      "RangeIndex: 388 entries, 0 to 387\n",
      "Data columns (total 12 columns):\n",
      " #   Column          Non-Null Count  Dtype  \n",
      "---  ------          --------------  -----  \n",
      " 0   Cohort          388 non-null    object \n",
      " 1   Individual_ID   388 non-null    object \n",
      " 2   Biological_Sex  388 non-null    object \n",
      " 3   Age_death       388 non-null    object \n",
      " 4   Disorder        388 non-null    object \n",
      " 5   1000G_ancestry  388 non-null    object \n",
      " 6   Genotype_data   388 non-null    object \n",
      " 7   snATAC-Seq      55 non-null     object \n",
      " 8   PMI             325 non-null    float64\n",
      " 9   pH              99 non-null     float64\n",
      " 10  RIN             66 non-null     float64\n",
      " 11  Notes           4 non-null      object \n",
      "dtypes: float64(3), object(9)\n",
      "memory usage: 36.5+ KB\n"
     ]
    }
   ],
   "source": [
    "df.info()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "2f10b6a3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Cohort</th>\n",
       "      <th>Individual_ID</th>\n",
       "      <th>Biological_Sex</th>\n",
       "      <th>Age_death</th>\n",
       "      <th>Disorder</th>\n",
       "      <th>1000G_ancestry</th>\n",
       "      <th>Genotype_data</th>\n",
       "      <th>snATAC-Seq</th>\n",
       "      <th>PMI</th>\n",
       "      <th>pH</th>\n",
       "      <th>RIN</th>\n",
       "      <th>Notes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>count</th>\n",
       "      <td>388</td>\n",
       "      <td>388</td>\n",
       "      <td>388</td>\n",
       "      <td>388</td>\n",
       "      <td>388</td>\n",
       "      <td>388</td>\n",
       "      <td>388</td>\n",
       "      <td>55</td>\n",
       "      <td>325.000000</td>\n",
       "      <td>99.000000</td>\n",
       "      <td>66.000000</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>unique</th>\n",
       "      <td>12</td>\n",
       "      <td>388</td>\n",
       "      <td>2</td>\n",
       "      <td>84</td>\n",
       "      <td>10</td>\n",
       "      <td>6</td>\n",
       "      <td>4</td>\n",
       "      <td>2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>top</th>\n",
       "      <td>CMC</td>\n",
       "      <td>CMC_MSSM_002</td>\n",
       "      <td>male</td>\n",
       "      <td>89+</td>\n",
       "      <td>control</td>\n",
       "      <td>EUR</td>\n",
       "      <td>Array</td>\n",
       "      <td>snMultiome</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Duplicates:IsoHuB:HSB189;Ma_et_al:HSB189</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>freq</th>\n",
       "      <td>100</td>\n",
       "      <td>1</td>\n",
       "      <td>252</td>\n",
       "      <td>32</td>\n",
       "      <td>166</td>\n",
       "      <td>316</td>\n",
       "      <td>268</td>\n",
       "      <td>40</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mean</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>19.814462</td>\n",
       "      <td>6.741414</td>\n",
       "      <td>7.006061</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>std</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>13.520554</td>\n",
       "      <td>3.651782</td>\n",
       "      <td>1.500808</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>min</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.300000</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>3.000000</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25%</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>10.800000</td>\n",
       "      <td>6.200000</td>\n",
       "      <td>6.325000</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50%</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>17.700000</td>\n",
       "      <td>6.400000</td>\n",
       "      <td>7.400000</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75%</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>25.000000</td>\n",
       "      <td>6.700000</td>\n",
       "      <td>8.225000</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>max</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>80.300000</td>\n",
       "      <td>42.300000</td>\n",
       "      <td>8.900000</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       Cohort Individual_ID Biological_Sex Age_death Disorder 1000G_ancestry  \\\n",
       "count     388           388            388       388      388            388   \n",
       "unique     12           388              2        84       10              6   \n",
       "top       CMC  CMC_MSSM_002           male       89+  control            EUR   \n",
       "freq      100             1            252        32      166            316   \n",
       "mean      NaN           NaN            NaN       NaN      NaN            NaN   \n",
       "std       NaN           NaN            NaN       NaN      NaN            NaN   \n",
       "min       NaN           NaN            NaN       NaN      NaN            NaN   \n",
       "25%       NaN           NaN            NaN       NaN      NaN            NaN   \n",
       "50%       NaN           NaN            NaN       NaN      NaN            NaN   \n",
       "75%       NaN           NaN            NaN       NaN      NaN            NaN   \n",
       "max       NaN           NaN            NaN       NaN      NaN            NaN   \n",
       "\n",
       "       Genotype_data  snATAC-Seq         PMI         pH        RIN  \\\n",
       "count            388          55  325.000000  99.000000  66.000000   \n",
       "unique             4           2         NaN        NaN        NaN   \n",
       "top            Array  snMultiome         NaN        NaN        NaN   \n",
       "freq             268          40         NaN        NaN        NaN   \n",
       "mean             NaN         NaN   19.814462   6.741414   7.006061   \n",
       "std              NaN         NaN   13.520554   3.651782   1.500808   \n",
       "min              NaN         NaN    1.300000   2.000000   3.000000   \n",
       "25%              NaN         NaN   10.800000   6.200000   6.325000   \n",
       "50%              NaN         NaN   17.700000   6.400000   7.400000   \n",
       "75%              NaN         NaN   25.000000   6.700000   8.225000   \n",
       "max              NaN         NaN   80.300000  42.300000   8.900000   \n",
       "\n",
       "                                           Notes  \n",
       "count                                          4  \n",
       "unique                                         4  \n",
       "top     Duplicates:IsoHuB:HSB189;Ma_et_al:HSB189  \n",
       "freq                                           1  \n",
       "mean                                         NaN  \n",
       "std                                          NaN  \n",
       "min                                          NaN  \n",
       "25%                                          NaN  \n",
       "50%                                          NaN  \n",
       "75%                                          NaN  \n",
       "max                                          NaN  "
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# that is better ... so, what is in this file?\n",
    "# https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.describe.html\n",
    "df.describe(include='all')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "31080819",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
